src/storage/storage.cpp

```diff
@@ -118,113 +73,103 @@ Storage::ReturnCode Storage::Run(int max_wait)
-    const SharedDataType data_region = regions_layout.old_data_region;
-
-    if (max_wait > 0)
+    // Get the next region ID and time stamp without locking shared barriers.
+    // Because of datastore_lock the only write operation can occur sequentially later.
+    auto next_region = REGION_1;
+    unsigned next_timestamp = 1;
+    auto in_use_region = REGION_NONE;
+    unsigned in_use_timestamp = 0;
+    if (SharedMemory::RegionExists(CURRENT_REGION))
     {
-        util::Log() << "Waiting for " << max_wait
-                    << " second for all queries on the old dataset to finish:";
+        auto shared_memory = makeSharedMemory(CURRENT_REGION);
+        auto current_region = static_cast<SharedDataTimestamp *>(shared_memory->Ptr());
+        in_use_region = current_region->region;
+        in_use_timestamp = current_region->timestamp;
+        next_region = in_use_region == REGION_1 ? REGION_2 : REGION_1;
+        next_timestamp = in_use_timestamp + 1;
     }
-    else
+
+    // ensure that the shared memory region we want to write to is really removed
+    // this is only needef for failure recovery because we actually wait for all clients
+    // to detach at the end of the function
+    if (storage::SharedMemory::RegionExists(next_region))
     {
-        util::Log() << "Waiting for all queries on the old dataset to finish:";
-    }
+        util::Log(logWARNING) << "Old shared memory region " << regionToString(next_region) << " still exists.";
+        util::UnbufferedLog() << "Retrying removal... ";
 
-    boost::interprocess::scoped_lock<boost::interprocess::named_sharable_mutex> regions_lock(
-        regions_layout.old_region_mutex, boost::interprocess::defer_lock);
+        // aquire a handle for the next shared memory region before we mark it for deletion
+        // we will need this to wait for all users to detach
+        auto next_shared_memory = makeSharedMemory(next_region);
 
-    if (max_wait > 0)
-    {
-        if (!regions_lock.timed_lock(boost::posix_time::microsec_clock::universal_time() +
-                                     boost::posix_time::seconds(max_wait)))
-        {
-            util::Log(logWARNING) << "Queries did not finish in " << max_wait
-                                  << " seconds. Claiming the lock by force.";
-            // WARNING: if queries are still using the old dataset they might crash
-            if (regions_layout.old_data_region == REGION_1)
-            {
-                barriers.resetRegion1();
-            }
-            else
-            {
-                BOOST_ASSERT(regions_layout.old_data_region == REGION_2);
-                barriers.resetRegion2();
-            }
+        storage::SharedMemory::Remove(next_region);
+        util::UnbufferedLog() << "ok.";
 
-            return ReturnCode::Retry;
-        }
+        util::UnbufferedLog() << "Waiting for clients to detach... ";
+        next_shared_memory->WaitForDetach();
```

Comment:
Sorry, i don't understand how `mlock` is related here. The problem what i see is that `osrm-datastore` is waiting for an event `all routed requests using this memory block are terminated` that may never happen. On other side it is not a problem to remove outdated block that can be still in use. This will unassign ftok key ID of that block. Than it is possible allocate a new block and assign the key ID to the newly allocated block. 

For sure an outdated block will not be deallocated if any request still holds facade that was attached to the block. But it should solved either by restarting `osrm-routed` or by introducing a request timeout like 5 minutes.